lib/db.js

```diff
@@ -1114,6 +1122,86 @@ module.exports = (
       })
   }
 
+  // Reduce redis memory usage by not encoding the keys. Store properties
+  // as fixed indices into arrays instead. Takes an unpacked session tokens
+  // structure as its argument, returns an array.
+  function packTokensForRedis (tokens) {
+    return Object.keys(tokens).reduce((result, tokenId) => {
+      const unpackedToken = tokens[tokenId]
+
+      result[tokenId] = truncatePackedArray(REDIS_SESSION_TOKEN_PROPERTIES.map(
+        (property, index) => {
+          const value = unpackedToken[property]
+
+          if (index === REDIS_SESSION_TOKEN_LOCATION_INDEX && value) {
+            return truncatePackedArray(REDIS_SESSION_TOKEN_LOCATION_PROPERTIES.map(
+              locationProperty => value[locationProperty]
+            ))
+          }
+
+          return unpackedToken[property]
+        }
+      ))
+
+      return result
+    }, {})
+  }
+
+  // Trailing null and undefined don't need to be stored.
+  function truncatePackedArray (array) {
+    const length = array.length
+    if (length === 0) {
+      return array
+    }
+
+    const item = array[length - 1]
+    if (item !== null && item !== undefined) {
+      return array
+    }
+
+    array.pop()
```

Comment:
Any value in making this a `while` loop to pop multiple undefined entries from the end of the array?

Reply:
It recurs below, so it is popping multiple items. I implemented it as a loop initially as it happens, but it seemed a little bit ugly. Recursion felt more elegant I guess, although that's obviously subjective and evidently less clear when reading the code. I can change it to a loop if that's preferred.

Reply:
No preference, just an idle thought